### PR TITLE
Move mouse event disabling on interactive elements to SimpleEventPlugin. Related perf tweak to click handlers.

### PR DESCRIPTION
**Pull request by nhunzaker**
Eliminates `DisabledInputUtils` in favor of filtering out mouse events for disabled interactive elements in `SimpleEventPlugin`.  

I've published the build of this branch, with a test page here: http://natehunzaker.com/disabled-react-inputs/

**Drawback (and a question):**

`ReactTestUtils.Simulate` now allows disabled inputs to receive mouse events. This made it necessary to use `ReactTestUtils.SimulateNative` when I updated the tests. I'm not sure what to do about this. It's as if `React.TestUtils.Simulate` is bypassing `SimpleEventPlugin`. Is this expected?

**Click bubbling performance tweak:**

I noticed that there's a fix for click event bubbling in iOS. iOS doesn't allow click bubbling for non-interactive elements, so a workaround is made by adding a local listener ([explained here](http://www.quirksmode.org/blog/archives/2010/09/click_event_del.html)). 

But now there's a check for interactive elements! So this fix can be filtered down to non-interactive elements, significantly reducing listener attachments.

[I've summarized this, and included Chrome JS Profiler snapshots here.](https://gist.github.com/nhunzaker/020335ee8ec0f1d487f97c7ddcfbd4a2)

---

**More background:**

A while back I submitted a fix for disabled inputs in IE (#6215). At the time, I believed that the fix could happen closer to the event system, but lacked familiarity with it.

#7616 yielded [a video deep dive on the event system](https://www.youtube.com/watch?v=dRo_egw7tBc), which was super helpful in understanding this part of React. Thanks kentcdodds, gaearon,  and spicyj (not \@mentioned for noise)!

### CODE REVIEW

File: src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js [hunk 1]
```diff
@@ -634,7 +656,7 @@ var SimpleEventPlugin = {
   },
 
   willDeleteListener: function(inst, registrationName) {
-    if (registrationName === 'onClick') {
+    if (registrationName === 'onClick' && isInteractive(inst._tag)) {
```
Comment (Aweary) [inline src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js]: Isn't this the opposite of what we want? i.e., we want to add the empty click handler to elements that aren't considered interactive?

**Comment by Aweary** [comment]
I like the idea of removing a level of indirection here, but if it breaks `ReactTestUtils.Simulate` then we'd need to address that somehow.

File: src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js (hunk 1 above)
Comment (nhunzaker) [inline src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js]: Oof... Yes. Looks like there's also no test coverage here. I'll add that as well.

**Comment by Aweary** [comment]
cc @spicyj

File: src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js (hunk 1 above)
Comment (nhunzaker) [inline src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js]: Done. I inverted the check and added test cases.

Comment (spicyj) [inline src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js]: I don't understand what this change is for. Can you explain?

**Comment by nhunzaker** [comment]
@Aweary Did a bit more research. I modified the `ReactTestUtils.Simulate.*` such that it uses the event hub. **With the exception of change events**, it works!

Checkout: https://github.com/nhunzaker/react/compare/nh-disabled-mouse-extraction...nhunzaker:nh-disabled-mouse-extraction-simulation

As for `change` events... 

The trouble is that, when `EventHubPlugin` calls out to `ChangeEventPlugin`, it absorbs the `onChange` event because `inputValueTracking.js` deems the value has not changed. I'm observing it in the following test (and others that are similar):

```javascript
// ReactDOMInput-test.js:248
it('should properly control a value of number `0`', function() {
  var stub = <input type="text" value={0} onChange={emptyFunction} />;
  stub = ReactTestUtils.renderIntoDocument(stub);
  var node = ReactDOM.findDOMNode(stub);

  node.value = 'giraffe';
  ReactTestUtils.Simulate.change(node);
  expect(node.value).toBe('0');
});
```

Basically, `onChange` is getting ignored because [`inputValueTracking` overrides the `value` property and observes new assignments](https://github.com/facebook/react/blob/master/src/renderers/dom/client/inputValueTracking.js#L80-L90). So when the change event fires afterwards, technically _nothing has changed_, and `ChangeEventPlugin` ignores it.

So far I'm exploring 2 options:

1. If `EventPluginHub` doesn't pick up any events, just dispatch a SyntheticEvent. It's already whitelisted to trusted event types.
2. If the event is a `change` event, use the current Simulate behavior of always dispatching a SyntheticEvent (which circumvents the event hub and forces the `onChange` handler).

What do you think?

File: src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js (hunk 1 above)
Comment (nhunzaker) [inline src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js]: I had it reversed (sorry). It now reads `if (registrationName === 'onClick'
&& !isInteractive(inst._tag)) {`, inverting the isInteractive check.

Basically, iOS safari only exhibits the event bubbling issue for
non-interactive elements (like divs). Extra event handlers can be
eliminated by only attaching this event listener on those non-interactive
elements. Since the check was already created for the mouse events filter,
I went ahead and added it here.

This trades an event listener attachment for function invocation (that'll
probably get inlined), when I ran the numbers it seemed much faster
https://gist.github.com/nhunzaker/020335ee8ec0f1d487f97c7ddcfbd4a2.

**Comment by spicyj** [comment]
GH line comments are broken, seemingly so I'll just write here.

I might be okay with breaking Simulate in this case actually. Need to think more.

This suppresses for contextmenu, mouseover, mouseout too which the original didn't. Is that intentional? If it is intentional for over/out then we should probably do enter/leave too right? If we want this before the next major maybe we can limit it to the 5 events that are currently suppressed.

When is `inst._currentElement` null (in `inst._currentElement &&`)?

Can you reformat the isInteractive to be

```
return (
  tag === 'button' || tag === 'input' ||
  tag === 'select' || tag === 'textarea'
);
```

or a switch statement returning true/false?

**Comment by Aweary** [comment]
@spicyj did you mean to mark as accepted? Seems like there's still potential changes to make

**Comment by nhunzaker** [comment]
> I might be okay with breaking Simulate in this case actually. Need to think more.

I'm starting to think this is an okay behavior. If I tap into `EventPluginHub`, I get other test failures related to the other event plugins. For example, mouse enter/leave fail. `Simulate` _forcing_ an event, with `SimulateNative` producing the expected DOM behavior, is okay by me (for whatever my vote is worth).

> This suppresses for contextmenu, mouseover, mouseout. Is that intentional? [...]

I've reordered the event switch to rule out these cases, similarly to the exceptions for Firefox. Though I'm unclear if the other events fire on `disabled`, I'd be happy to look into it. Either way, I agree that it is important for a non-breaking release.

> When is inst._currentElement null (in inst._currentElement &&)?

It otherwise fails the following test: 

https://github.com/facebook/react/blob/master/src/renderers/dom/client/__tests__/ReactEventIndependence-test.js#L53-L68

Though I'm curious if this could just be an invariant in `Simulate`.

> Can you reformat the isInteractive to be [...] or a switch statement returning true/false?

Yep. You know I wrote a `switch` twice before I settled on the `||` chain... I've done the reformatting.

**Comment by nhunzaker** [comment]
Sort of unrelated, but when I was working through an attempt to update `Simulate`, I noticed that [BeforeInputEventPlugin returns `[null, null]` a lot](https://github.com/facebook/react/blob/master/src/renderers/dom/client/eventPlugins/BeforeInputEventPlugin.js#L460-L473). This happens, I think, nearly every time `EventPluginHub.extractEvents` is invoked. Any interest in a PR that cuts the extra array allocation?

**Comment by nhunzaker** [comment]
@Aweary I think I've addressed all of the comments. The only thing left is to figure out how the change to `Simulate` should be handled.

**Comment by Aweary** [comment]
Thanks!
> The only thing left is to figure out how the change to Simulate should be handled.

I'm going to defer to @spicyj on that

File: src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js [hunk 2]
```diff
@@ -137,6 +137,23 @@ function getDictionaryKey(inst) {
   return '.' + inst._rootNodeID;
 }
 
+function isInteractive(tag) {
+  return tag === 'button' || tag === 'input' ||
```
Comment (spicyj) [inline src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js]:
```
return (
  tag === 'button' || tag === 'input' ||
  tag === 'select' || tag === 'textarea'
);
```

FB style doesn't do large indents to line up with previous lines – we just move the previous line down instead.

Comment (nhunzaker) [inline src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js]: 👍

**Comment by spicyj** [comment]
I'm fine with the simulate change for a minor release. Can you fix up the style and rebase? Otherwise this looks great, thank you!

File: src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js (hunk 2 above)
Comment (nhunzaker) [inline src/renderers/dom/client/eventPlugins/SimpleEventPlugin.js]: Sorry, for whatever reason I totally caught your comment wrong before. Just a sec.

**Comment by nhunzaker** [comment]
@spicyj Done.

**Comment by spicyj** [comment]
Thank you!
